value/src/main/java/com/google/auto/value/AutoValueExtension.java

```diff
@@ -0,0 +1,33 @@
+package com.google.auto.value;
+
+import javax.annotation.processing.ProcessingEnvironment;
+import javax.lang.model.element.ExecutableElement;
+import javax.lang.model.element.TypeElement;
+import java.util.Collection;
+import java.util.Map;
+
+/**
+ * Created by rharter on 5/1/15.
+ */
+public interface AutoValueExtension {
+
+  public interface Context {
+    ProcessingEnvironment processingEnvironment();
+    String packageName();
+    TypeElement autoValueClass();
+    Map<String, ExecutableElement> properties();
+  }
+
+  public interface GeneratedClass {
+    String className();
+    String source();
+    Collection<String> additionalImports();
+    Collection<ExecutableElement> consumedProperties();
+    Collection<String> additionalInterfaces();
+    Collection<String> additionalCode();
```

Comment:
It's important for extensions to be subclasses of the generated AutoValue implementation, so they can override the generated toString() etc. One of our use cases is exactly that (being able to omit certain properties from toString()).

There's nothing stopping extensions from having static properties or methods even if they are not at the bottom of the inheritance hierarchy. The issue with Parcelable seems to be that the CREATOR field has to be in the Parcelable class itself and can't be inherited from an ancestor. I personally think that design is wrong, but there it is. We're certainly [not the first](http://stackoverflow.com/questions/12809928/parcelable-and-inheritance) to encounter it.

So in conclusion I think we probably want a way for an extension to declare that it must be at the bottom of the hierarchy? If you have two such extensions and they both want to generate code for a given class then we'll just have to reject that.

The alternative design that we considered was to switch the AutoValue implementation so it uses [JavaPoet](https://github.com/square/javapoet) instead of templates, and have extensions be able to contribute code to the single generated class by calling JavaPoet methods. I've been reluctant to do that for a number of reasons. First, I like templates better than code-generating-code. Second, it doesn't give a clean solution for the toString() override. Third, it feels kind of anarchic when there is more than one extension contributing code (especially if they can change the already-contributed code). Finally, it risks having extensions know too much about exactly what the AutoValue implementation class looks like.

Despite all that, we may conclude from this exercise that the JavaPoet approach is the way to go.